### PR TITLE
ARROW-11254: [Rust][DataFusion] Add SIMD and snmalloc flags as options to benchmarks

### PR DESCRIPTION
**Pull request by Dandandan**
I think it would be great to test some compilation options individually on the benchmarks to see the impact.

Here are some examples of the impact it can have on the queries:

--features "":
```
Query 5 iteration 0 took 655.3 ms
Query 5 iteration 1 took 648.9 ms
Query 5 iteration 2 took 640.3 ms
Query 5 iteration 3 took 658.7 ms
Query 5 iteration 4 took 646.3 ms
Query 5 iteration 5 took 684.1 ms
Query 5 iteration 6 took 642.8 ms
Query 5 iteration 7 took 656.9 ms
Query 5 iteration 8 took 646.0 ms
Query 5 iteration 9 took 669.1 ms
Query 5 avg time: 654.85 ms
```

--features "snmalloc"

```
Query 5 iteration 0 took 525.4 ms
Query 5 iteration 1 took 478.8 ms
Query 5 iteration 2 took 485.7 ms
Query 5 iteration 3 took 486.6 ms
Query 5 iteration 4 took 482.6 ms
Query 5 iteration 5 took 473.1 ms
Query 5 iteration 6 took 494.4 ms
Query 5 iteration 7 took 483.5 ms
Query 5 iteration 8 took 493.1 ms
Query 5 iteration 9 took 479.4 ms
Query 5 avg time: 488.26 ms
```

--features ""
```
Query 12 iteration 0 took 241.4 ms
Query 12 iteration 1 took 234.8 ms
Query 12 iteration 2 took 229.8 ms
Query 12 iteration 3 took 229.5 ms
Query 12 iteration 4 took 228.3 ms
Query 12 iteration 5 took 230.0 ms
Query 12 iteration 6 took 228.3 ms
Query 12 iteration 7 took 229.3 ms
Query 12 iteration 8 took 229.9 ms
Query 12 iteration 9 took 230.1 ms
Query 12 avg time: 231.13 ms
```
--features "simd"
```
Query 12 iteration 0 took 157.7 ms
Query 12 iteration 1 took 159.3 ms
Query 12 iteration 2 took 156.9 ms
Query 12 iteration 3 took 163.0 ms
Query 12 iteration 4 took 157.5 ms
Query 12 iteration 5 took 157.6 ms
Query 12 iteration 6 took 156.6 ms
Query 12 iteration 7 took 157.4 ms
Query 12 iteration 8 took 158.6 ms
Query 12 iteration 9 took 157.0 ms
Query 12 avg time: 158.16 ms
```

### CODE REVIEW

**Comment by seddonm1** [comment]
Cool. Maybe you could add the instructions to the readme?

**Comment by Dandandan** [comment]
A further option that can have bigger impact is by adding 
```
[profile.release]
lto = true
```

to the project using DataFusion. But I am not sure if we can conditionally enable this (also it makes the build times very slow, but I think it makes sense to use for distributed binaries).

**Comment by seddonm1** [comment]
> A further option that can have bigger impact is by adding
> 
> ```
> [profile.release]
> lto = true
> ```
> 
> to the project using DataFusion. But I am not sure if we can conditionally enable this (also it makes the build times very slow, but I think it makes sense to use for distributed binaries).

I just read about this optimisation. What sort of performance are you seeing with it enabled (and simd)?

**Comment by Dandandan** [comment]
@seddonm1 

Query 12, with lto + simd   + snmalloc:

```
Query 12 iteration 0 took 118.8 ms
Query 12 iteration 1 took 118.7 ms
Query 12 iteration 2 took 118.3 ms
Query 12 iteration 3 took 117.0 ms
Query 12 iteration 4 took 117.7 ms
Query 12 iteration 5 took 118.9 ms
Query 12 iteration 6 took 117.9 ms
Query 12 iteration 7 took 118.1 ms
Query 12 iteration 8 took 117.7 ms
Query 12 iteration 9 took 117.8 ms
Query 12 avg time: 118.09 ms
```

**Comment by seddonm1** [comment]
Wow 654.85 -> 118.09 is huge.

**Comment by Dandandan** [comment]
> Wow 654.85 -> 118.09 is huge.
No unfortunately Q12 runs finishes ~230ms on master. But still 230 -> 118 is a factor 1.95 :)

**Comment by Dandandan** [comment]
For completeness, here is query 5 with simd + snmalloc + lto enabled:

```
Query 5 iteration 0 took 460.2 ms
Query 5 iteration 1 took 439.2 ms
Query 5 iteration 2 took 436.8 ms
Query 5 iteration 3 took 425.2 ms
Query 5 iteration 4 took 434.0 ms
Query 5 iteration 5 took 429.6 ms
Query 5 iteration 6 took 430.3 ms
Query 5 iteration 7 took 422.6 ms
Query 5 iteration 8 took 423.8 ms
Query 5 iteration 9 took 424.2 ms
Query 5 avg time: 432.59 ms
```

**Comment by seddonm1** [comment]
> > Wow 654.85 -> 118.09 is huge.
> 
> No unfortunately, Q12 runs finishes ~230ms on master. But still 230 -> 118 is a factor 1.95 :)

Oh sorry I miss read it. Yes, I will happily take 230 -> 118 especially as the current option is Spark which hasn't even started the JVM for multiple seconds after this!

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Thanks a lot for sharing the snmalloc. Looks very significant for our use-case!

File: rust/benchmarks/Cargo.toml [hunk 1]
```diff
@@ -26,11 +26,16 @@ repository = "https://github.com/apache/arrow"
...
 arrow = { path = "../arrow" }
 parquet = { path = "../parquet" }
 datafusion = { path = "../datafusion" }
 structopt = { version = "0.3", default-features = false }
 tokio = { version = "0.2", features = ["macros", "rt-core", "rt-threaded"] }
 futures = "0.3"
-env_logger = "^0.8"
\ No newline at end of file
+env_logger = "^0.8"
+snmalloc-rs = {version = "0.2", optional = true }
```
Comment (jorgecarleitao) [inline rust/benchmarks/Cargo.toml:41]: nit: add line ending

**Comment by alamb** [comment]
I apologize for the delay in merging Rust PRs -- the 3.0 release is being finalized now and are planning to minimize entropy by postponing merging  changes not critical for the release until the process was complete. I hope the process is complete in the next few days. There is more [discussion](https://lists.apache.org/list.html?dev@arrow.apache.org) in the mailing list

**Pull request closed without merging**